shard/sharder.go

```diff
@@ -0,0 +1,352 @@
+	err = s.rpcClient.Call(targetPeer, "Cluster", "IPFSBlockPut",
+		b, &retStr)
+	if err != nil {
+		return err
+	}
+
+	// Track shard node as a normal pin within cluster state
+	pinS := api.Pin{
+		Cid:                  shardNode.Cid(),
+		Allocations:          []peer.ID{targetPeer},
+		ReplicationFactorMin: 1,
+		ReplicationFactorMax: 1,
+	}.ToSerial()
+
+	return s.rpcClient.Call(targetPeer, "Cluster", "Pin", pinS, &struct{}{})
+}
+
+/* Open questions
+
+1. getAssignment()
+   How to decide how much space to give out?  There seems to be an infinity
+   of different ways to do this so ultimately I see this being configurable
+   perhaps keeping strategies in the allocator config.  Orthoganal to this
+   is deciding WHICH peer gets the next shard.  Currently this is the choice
+   the cluster pin allocator makes and of course there is a rabbit hole of
+   work to get different properties and guarantees from this decision too.
+   Perhaps these two decisions will be made as part of a more general
+   decision incorporating other things as well?
+
+   For now I am assuming the WHICH peer decision is made first (probably
+   the peer with most open disk space) and want a way to answer HOW MUCH
+   space do we take from this peer to store a shard?  A few basic ideas:
+
+   A. All nodes offer up some fraction of their space (say 2/3) during each
+      allocation.  The emptiest nodes are allocated first. The second
+      alloc gets 2/3 of remaining 1/3 = 2/6 of the total empty disk space
+      etc.
+   B. Same as A but saturates in a finite number of steps.  First allocate
+      2/3 of remaining. If peer is allocated again allocate remaining 1/3
+   C. All nodes offer up min space that any node can afford.  All nodes
+      are allocated before repeating allocations (this bleeds into the
+      WHICH node decision)
+
+  Note pseudo-code as it stands doesn't take into account the shard node
+  when keeping block put data within threshold.  Size can't be determined
+  beforehand because blocks may be of different sizes.
+
+  We will probably be reusing the pin allocator.  Where is the best place for
+  the HOW MUCH decision code, generally in an allocator or here in the
+  sharder responding to information received from informer calls?
+
+2. Shard node (cluster dag) format
+   In the current go-ipfs landscape downloading resolvers for non-standard
+   ipld formats (i.e. everything but unixfs, and I think dag-cbor) is
+   a difficult requirement to ask from nodes in the cluster.
+
+   For this reason I think we should use one of ipfs's standard ipld
+   formats.  Is dag-cbor ok to use? Probably preferrable over unixfs
+   protobuf if we can help it.
+
+3. We'll need to think about serializing nodes across rpc.  This may be
+   simple I need to dig into this.  Depending on the node format that
+   files are imported to we might need multiple api types for multiple
+   formats.  We might be able to guarantee one format from our importer
```

Comment:
From the moment the data is chunked and we serialize it to pass it to the sharder, I don't see there is need to deserialize it anymore. We just need to carry the right metadata and block-put the binary slice we are carrying. I think this is how you do it now for the most part.

Reply:
Thanks for noticing this